### DISCUSSION TITLE
gdsfactory sweepme plugin

### DISCUSSION THREAD

**Discussion by joamatab** [SweepMe/pysweepme#21] [Q&A]
How could we build a [gdsfactory](https://github.com/gdsfactory/gdsfactory) plugin for sweepme

Ideally as we do the design we can specify the measurement sequence

Also, Where can we find the code for the instrument drivers?
https://sweep-me.net/devices

I noticed that you have many other projects listed here
https://sweep-me.net/python_projects

**Comment by afischer-sweepme** [comment]
Hi @joamatab,

as I do not know much about gdsfactory I cannot say what would be the best the way to create some kind of interface. It also depends on whether you aim at using pysweepme or SweepMe!. Here, I would suggest to meet online.

In general, the field of test&measurement has the problem that each user has a different set of instruments which makes it difficult to create "the" measurement sequence for all. SweepMe! already aims at resolving to some extent this by introducing a Module layer that is instrument independent and where each user can select different instrument drivers in the Module to customize a program procedure.

You can access the code of our instrument drivers as explained here https://pypi.org/project/pysweepme/ in the section "Instrument drivers". We aim at creating a new github repository just for drivers so that it will become easier to contribute. Still, the versioned instrument drivers are only available via download from the webpage or in SweepMe! using the version manager.

The other python projects that we are listing on our webpage should help to get an overview about all those python activities about measurement automation. There are quite a lot, but standardization is missing so far. It means that you cannot easily combine instrument drivers from different projects. Still, all these projects are a great resource for learning how to interface with measurement instruments. Also more and more instrument manufactureres start to publish their official python libraries that we aim to present there.

Best, Axel

**Comment by joamatab** [comment]
Hi Axel,

We would like to leverage your abstraction layer for instruments to be able to create a test sequence standard, using YAML that we can generate from gdsfactory

GitHub repo with drivers would be great

@HelgeGehring 
@simbilod 
@mdecea

**Comment by afischer-sweepme** [comment]
Hi, we plan the release of the Drivers on github in Q2 in 2023. The download of drivers from our website will be still the recommended way in future to deploy drivers in own project as each driver get a version when uploaded to our server. Instead of loading all drivers by importing an entire package, we use pysweepme to make it possible to load exactly those drivers as needed. This will be important when there is a growing number of drivers and when one likes to change a driver version without changing other drivers at the same time. SweepMe! itself also uses pysweepme to load drivers from specified folders.

You can also have a look at the repo from @gtortone who created and contributed several well-written drivers that have been made available via our server afterwards: https://github.com/gtortone/sweepme-drivers

SweepMe! instrument drivers now have a double use: They can be used with the SweepMe! software to create test sequences by draf&drop of modules. For this reason, one needs to fill semantic function like 'connect', 'initialize', 'apply', 'measure' and so on. At the same time we also have started with the release of pysweepme to wrap communication commands in convenience function like 'set_voltage', or 'get_current' which are device dependent (in contrast to the semantic functions). This allows to load SweepMe! instrument drivers with pysweepme in own python projects and call communication commands easily.  

The semantic functions could become interesting for you again when you like to adopt our abstraction layer with a module that represents some kind of instrument like wafer prober, parameter analyzer, source-measuring unit, etc. Not all instrument have exactly the same communication commands and therefore the semantic functions can be helpful to do some meaningful in each phase of a measurement procedure without knowing which instrument is actually used.
